File: backend/health_vault/doctor_visit.py
"""Doctor Visit Mode — printable professional summary."""

from __future__ import annotations

from typing import Any

from backend.health_vault.models import UserDashboardPreferences, utc_now
from backend.health_vault.timeline import build_timeline
from backend.health_vault.unit_conversion import CANONICAL_UNITS, apply_display_units
from backend.health_vault.vault_store import VaultStore


class DoctorVisitMode:
    def __init__(self, store: VaultStore) -> None:
        self.store = store

    def _prefs(self, patient_id: str) -> UserDashboardPreferences:
        profile = self.store.get_profile(patient_id=patient_id) or {}
        return UserDashboardPreferences.from_dict(profile.get("dashboard_preferences") or {})

    def _trend_line(self, metric: str, patient_id: str, *, label: str | None = None) -> str:
        t = (self.store.get_trends(patient_id=patient_id) or {}).get(metric) or {}
        if not t:
            return "n/a"
        latest = t.get("latest")
        direction = t.get("label") or "n/a"
        prefs = self._prefs(patient_id)
        display = apply_display_units(
            {
                "metric": metric,
                "observation_class": metric,
                "value": latest,
                "units": CANONICAL_UNITS.get(metric),
            },
            region=prefs.reporting_region,
            unit_overrides=prefs.unit_overrides,
        )
        shown = display.get("display_value")
        units = display.get("display_units") or ""
        if shown is None:
            shown = latest
        prefix = f"{label} " if label else ""
        if shown is not None:
            unit_bit = f" {units}" if units else ""
            return f"{prefix}{direction} (latest {shown}{unit_bit})".strip()
        return f"{prefix}{direction}".strip()

    def _diabetes_trend(self, patient_id: str) -> str:
        trends = self.store.get_trends(patient_id=patient_id) or {}
        glucose_keys = (
            "glucose_random",
            "glucose",
            "glucose_fasting",
            "glucose_cgm_interstitial",
            "glucose_postprandial",
            "glucose_capillary",
            "glucose_serum_plasma",
        )
        if not any(trends.get(key) for key in glucose_keys):
            return f"n/a · HbA1c {self._trend_line('hba1c', patient_id)}"
        parts: list[str] = []
        if trends.get("glucose_random"):
            parts.append(self._trend_line("glucose_random", patient_id, label="random"))
        elif trends.get("glucose"):
            parts.append(self._trend_line("glucose", patient_id, label="unspecified"))
        if trends.get("glucose_fasting"):
            parts.append(self._trend_line("glucose_fasting", patient_id, label="fasting"))
        if trends.get("glucose_cgm_interstitial"):
            parts.append(self._trend_line("glucose_cgm_interstitial", patient_id, label="CGM"))
        parts.append(f"HbA1c {self._trend_line('hba1c', patient_id)}")
        return " · ".join(parts) if parts else "n/a"

    def generate(self, patient_id: str = "default-patient") -> dict[str, Any]:
        profile = self.store.get_profile(patient_id=patient_id)
        if not profile and patient_id == "default-patient":
            profile = self.store.get_profile()
        docs = [
            document
            for document in self.store.list_documents()
            if document.get("patient_id", "default-patient") == patient_id
        ]
        timeline = build_timeline(self.store, patient_id=patient_id)
        ecg = [
            d
            for d in docs
            if d.get("document_type") == "samsung_health_ecg" or "ecg" in (d.get("tags") or [])
        ][-3:]
        kidney = self._trend_line("egfr", patient_id)
        return {
            "title": "HealthChecker+ Doctor Visit Report",
            "generated_at": utc_now(),
            "patient_id": patient_id,
            "fhir_bundle_hint": [
                "Patient",
                "Medication",
                "Observation",
                "DiagnosticReport",
                "DocumentReference",
                "Encounter",
            ],
            "current_diagnoses": profile.get("diagnoses") or [],
            "current_medications": profile.get("medications") or [],
            "recent_ecg": list(reversed(ecg)),
            "kidney_trend": kidney,
            "blood_pressure_trend": f"{self._trend_line('systolic_bp', patient_id)} / {self._trend_line('diastolic_bp', patient_id)}",
            "sleep_trend": self._trend_line("sleep_score", patient_id),
            "diabetes_trend": self._diabetes_trend(patient_id),
            "imported_reports": docs,
            "health_timeline": timeline[:25],
        }

File: js/health_vault/doctor_visit.js
/**
 * HC-201 — Doctor Visit Mode (printable professional summary).
 */
(function (global) {
  "use strict";

  function latestMeasurement(metric) {
    const items = global.HCHealthVault.listMeasurements({ metric }).slice();
    items.sort((a, b) =>
      String(b.measured_at || "").localeCompare(String(a.measured_at || ""))
    );
    return items[0] || null;
  }

  function trendLine(metric) {
    const t = (global.HCTrendEngine.getSnapshot() || {})[metric];
    if (!t) return "n/a";
    return t.label + (t.latest != null ? ` (latest ${t.latest})` : "");
  }

  function generateReport(options) {
    const opts = options || {};
    const Vault = global.HCHealthVault;
    const Timeline = global.HCHealthTimeline;
    const profile = Vault.getProfile();
    const docs = Vault.listDocuments();
    const timeline = Timeline.build();

    // Optionally merge legacy HC_V6 symptoms as soft context
    let legacy = null;
    try {
      if (typeof data === "function") legacy = data();
    } catch (_) {}

    const ecg = docs
      .filter((d) => d.document_type === "samsung_health_ecg" || (d.tags || []).includes("ecg"))
      .slice(-3)
      .reverse();

    const report = {
      title: "HealthChecker+ Doctor Visit Report",
      generated_at: new Date().toISOString(),
      patient_id: opts.patient_id || "default-patient",
      fhir_bundle_hint: [
        "Patient",
        "Medication",
        "Observation",
        "DiagnosticReport",
        "DocumentReference",
        "Encounter",
      ],
      current_diagnoses: profile.diagnoses || [],
      current_medications: profile.medications || [],
      recent_ecg: ecg.map((d) => ({
        id: d.id,
        filename: d.original_filename,
        imported_at: d.imported_at,
        storage_uri: d.storage_uri,
      })),
      kidney_trend: trendLine("egfr"),
      blood_pressure_trend: `${trendLine("systolic_bp")} / ${trendLine("diastolic_bp")}`,
      sleep_trend: trendLine("sleep_score"),
      diabetes_trend: `${trendLine("glucose")} · HbA1c ${trendLine("hba1c")}`,
      latest: {
        glucose: latestMeasurement("glucose"),
        egfr: latestMeasurement("egfr"),
        systolic: latestMeasurement("systolic_bp") || latestMeasurement("systolic"),
        diastolic: latestMeasurement("diastolic_bp") || latestMeasurement("diastolic"),
        sleep_score: latestMeasurement("sleep_score"),
      },
      imported_reports: docs.map((d) => ({
        id: d.id,
        type: d.document_type,
        filename: d.original_filename,
        imported_at: d.imported_at,
        confidence: d.parser_confidence,
        sha256: d.sha256,
        provenance:
          d.provenance ||
          ((d.tags || []).find((t) => String(t).indexOf("provenance:") === 0) || "")
            .toString()
            .replace(/^provenance:/, "") ||
          "unspecified",
      })),
      health_timeline: timeline.slice(0, 25),
      medical_disclaimer:
        "Observational decision-support only. Not a diagnosis. User-reported and historical-summary values are not laboratory-document verified.",
      legacy_summary: legacy
        ? {
            readings: (legacy.logs || []).length,
            symptoms: legacy.sym || [],
            foot_pain_analyses: (legacy.footPainLogs || []).length,
          }
        : null,
    };
    return report;
  }

  function renderPrintable(targetEl) {
    const report = generateReport();
    if (!targetEl) return report;

    const dx =
      report.current_diagnoses.length > 0
        ? report.current_diagnoses.map((x) => `<li>${esc(x)}</li>`).join("")
        : "<li class='muted'>None recorded in vault profile</li>";
    const meds =
      report.current_medications.length > 0
        ? report.current_medications.map((x) => `<li>${esc(x)}</li>`).join("")
        : "<li class='muted'>None recorded in vault profile</li>";
    const imports = report.imported_reports
      .slice()
      .reverse()
      .slice(0, 15)
      .map(
        (d) =>
          `<li>${esc(d.type)} — ${esc(d.filename || d.id)} <span class="small muted">${esc(
            d.provenance
          )} · ${esc(d.imported_at)}</span></li>`
      )
      .join("");

    const timeline = report.health_timeline
      .map(
        (e) =>
          `<li><strong>${esc(new Date(e.date).toLocaleDateString())}</strong> ${esc(
            e.document.document_type
          )} — ${esc(e.trend_impact)}</li>`
      )
      .join("");

    targetEl.innerHTML = `
      <div class="doctor-visit-report">
        <h3 class="section-title">${esc(report.title)}</h3>
        <div class="small muted">Generated ${esc(new Date(report.generated_at).toLocaleString())}</div>
        <div class="small muted">${esc(report.medical_disclaimer)}</div>
        <div class="hr"></div>
        <div class="kpi"><strong>Current diagnoses</strong><ul>${dx}</ul></div>
        <div class="kpi"><strong>Current medications</strong><ul>${meds}</ul></div>
        <div class="kpi"><strong>Recent ECG</strong>
          ${
            report.recent_ecg.length
              ? report.recent_ecg
                  .map((e) => `<div class="small">${esc(e.filename || e.id)} (${esc(e.imported_at)})</div>`)
                  .join("")
              : '<div class="muted small">No ECG imports yet</div>'
          }
        </div>
        <div class="kpi"><strong>Kidney trend:</strong> ${esc(report.kidney_trend)}</div>
        <div class="kpi"><strong>Blood pressure trend:</strong> ${esc(report.blood_pressure_trend)}</div>
        <div class="kpi"><strong>Sleep trend:</strong> ${esc(report.sleep_trend)}</div>
        <div class="kpi"><strong>Diabetes trend:</strong> ${esc(report.diabetes_trend)}</div>
        <div class="kpi"><strong>Imported reports</strong><ul>${imports || "<li class='muted'>None</li>"}</ul></div>
        <div class="kpi"><strong>Health timeline</strong><ul>${timeline || "<li class='muted'>Empty</li>"}</ul></div>
        <button type="button" onclick="window.print()">Print Doctor Visit Report</button>
      </div>
    `;
    return report;
  }

  function esc(v) {
    return String(v == null ? "" : v)
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;");
  }

  global.HCDoctorVisit = {
    generateReport,
    renderPrintable,
  };
})(typeof window !== "undefined" ? window : globalThis);
